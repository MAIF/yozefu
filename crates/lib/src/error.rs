//! Error definitions and their implementations of the `From` trait.
use rdkafka::error::KafkaError;
use std::fmt::Display;
use std::str::Utf8Error;
use std::{
    fmt::{self, Formatter},
    num::TryFromIntError,
};

/// All kinds of errors that may occur
#[derive(Debug)]
pub enum Error {
    Error(String),
    KafkaError(KafkaError),
    IoError(std::io::Error),
    SerdeError(serde_json::Error),
    ThemeError(String),
    Search(SearchError),
    SchemaRegistry(String),
    Tokio(String),
}

#[derive(Debug)]
pub enum SearchError {
    Parse(String),
}

impl std::error::Error for Error {}

impl Display for Error {
    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
        match self {
            Error::IoError(e) => write!(f, "IO Error: {}", e),
            Error::SerdeError(e) => write!(f, "Serde Error: {}", e),
            Error::ThemeError(e) => write!(f, "Theme Error: {}", e),
            Error::KafkaError(e) => write!(f, "Kafka Error: {}", e),
            Error::Error(e) => write!(f, "{}", e),
            Error::Tokio(e) => write!(f, "Tokio Error: {}", e),
            Error::Search(e) => write!(f, "{}", e),
            Error::SchemaRegistry(e) => write!(f, "Schema registry Error: {}", e),
        }
    }
}

impl Display for SearchError {
    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
        match self {
            SearchError::Parse(e) => write!(f, "Cannot parse the search query at '{}'", e),
        }
    }
}

impl From<std::fmt::Error> for Error {
    fn from(e: std::fmt::Error) -> Self {
        Error::Error(e.to_string())
    }
}

impl From<KafkaError> for Error {
    fn from(e: KafkaError) -> Self {
        Error::KafkaError(e)
    }
}

impl From<std::io::Error> for Error {
    fn from(e: std::io::Error) -> Self {
        Error::IoError(e)
    }
}

impl From<serde_json::Error> for Error {
    fn from(e: serde_json::Error) -> Self {
        Error::SerdeError(e)
    }
}

impl From<&str> for Error {
    fn from(e: &str) -> Self {
        Error::Error(e.to_string())
    }
}

impl From<TryFromIntError> for Error {
    fn from(e: TryFromIntError) -> Self {
        Error::Error(e.to_string())
    }
}

impl From<strum::ParseError> for Error {
    fn from(e: strum::ParseError) -> Self {
        Error::Error(e.to_string())
    }
}

impl From<Utf8Error> for Error {
    fn from(e: Utf8Error) -> Self {
        Error::Error(e.to_string())
    }
}
