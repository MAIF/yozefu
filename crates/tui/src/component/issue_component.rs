use ratatui::{
    Frame,
    layout::{Constraint, Flex, Layout, Rect},
    style::{Style, Stylize},
    text::Line,
    widgets::{Block, BorderType, Borders, Clear, Padding, Paragraph},
};

use crate::error::TuiError;

use super::{Component, ComponentName, State};

#[derive(Default)]
pub(crate) struct IssueComponent {}

impl Component for IssueComponent {
    fn id(&self) -> ComponentName {
        ComponentName::Help
    }

    fn draw(&mut self, f: &mut Frame<'_>, rect: Rect, state: &State) -> Result<(), TuiError> {
        let block = Block::default()
            .borders(Borders::ALL)
            .padding(Padding::symmetric(5, 0))
            .style(Style::default())
            .border_type(BorderType::Rounded);

        let block = self.make_block_focused(state, block);
        let paragraph = Paragraph::new(vec![
            Line::from(""),
            Line::from("ʕノ•ᴥ•ʔノ ︵ ┻━┻"),
            Line::from("Alt: Your mate flipping the desk").italic(),
            Line::from(""),
            Line::from("Are you struggling with the tool?"),
            Line::from("Leave us an issue so we can improve it:"),
            Line::from("https://github.com/MAIF/yozefu/issues").bold(),
            Line::from(""),
            Line::from("Press any key to close.").italic(),
        ]);

        let mut rect = rect;
        rect.width -= 2;
        rect.y += 1;

        let [area] = Layout::horizontal([Constraint::Length(54)])
            .flex(Flex::End)
            .areas(rect);
        let [area] = Layout::vertical([Constraint::Length(11)])
            .flex(Flex::Start)
            .areas(area);

        f.render_widget(Clear, area);
        f.render_widget(paragraph.block(block), area);

        Ok(())
    }
}
