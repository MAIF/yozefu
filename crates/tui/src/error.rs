use lib::Error;
use std::{
    fmt::{self, Display, Formatter},
    num::TryFromIntError,
};
use tokio::sync::mpsc::error::SendError;

/// Wrapper of around the `lib::Error` struct.
//#[derive(Debug)]
pub struct TuiError(Error);

impl<T> From<SendError<T>> for TuiError {
    fn from(e: SendError<T>) -> Self {
        TuiError(Error::Error(e.to_string()))
    }
}

impl Display for TuiError {
    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
        write!(f, "{}", self.0)
    }
}

use std::fmt::Debug;

impl Debug for TuiError {
    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
        write!(f, "{}", self.0)
    }
}

impl From<&str> for TuiError {
    fn from(e: &str) -> Self {
        TuiError(Error::Error(e.to_string()))
    }
}

impl From<serde_json::Error> for TuiError {
    fn from(e: serde_json::Error) -> Self {
        TuiError(Error::SerdeError(e))
    }
}

impl From<Error> for TuiError {
    fn from(e: Error) -> Self {
        TuiError(e)
    }
}

impl From<TuiError> for Error {
    fn from(val: TuiError) -> Self {
        val.0
    }
}

impl From<std::io::Error> for TuiError {
    fn from(e: std::io::Error) -> Self {
        TuiError(Error::IoError(e))
    }
}

impl From<TryFromIntError> for TuiError {
    fn from(e: TryFromIntError) -> Self {
        TuiError(Error::Error(e.to_string()))
    }
}
