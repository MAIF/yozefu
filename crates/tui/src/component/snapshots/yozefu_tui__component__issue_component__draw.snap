---
source: crates/tui/src/component/issue_component.rs
expression: terminal.backend()
---
"                                                            "
"    ┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┓  "
"    ┃                                                    ┃  "
"    ┃     ʕノ•ᴥ•ʔノ ︵ ┻━┻                               ┃  " Hidden by multi-width symbols: [(12, " "), (18, " "), (21, " ")]
"    ┃     Alt: Your mate flipping the desk               ┃  "
"    ┃                                                    ┃  "
"    ┃     Are you struggling with the tool?              ┃  "
"    ┃     Leave us an issue so we can improve it:        ┃  "
"    ┃     https://github.com/MAIF/yozefu/issues          ┃  "
"    ┃                                                    ┃  "
"    ┃     Press any key to close.                        ┃  "
"    ┗━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┛  "
"                                                            "
"                                                            "
"                                                            "
